src/Core/Resolvers/MySqlQueryBuilder.cs

```diff
@@ -280,26 +280,26 @@ private string MakeInsertSelections(SqlInsertStructure structure)
         private string MakeUpsertSelections(SqlUpsertQueryStructure structure)
         {
             List<string> selections = new();
+            Dictionary<string, string> insertColumnsToParamName = structure.InsertColumns.Zip(structure.Values, (colName, paramName)
+                => new { Key = colName, Value = paramName }).ToDictionary(kv => kv.Key, kv => kv.Value);
 
             List<LabelledColumn> fields = structure.OutputColumns;
 
-            int index = 0;
             foreach (LabelledColumn column in fields)
             {
                 string quotedColName = QuoteIdentifier(column.Label);
-
-                if (structure.InsertColumns.Contains(column.ColumnName))
+                ColumnDefinition columnDefinition = structure.GetColumnDefinition(column.ColumnName);
+                if (insertColumnsToParamName.TryGetValue(column.ColumnName, out string? paramName))
                 {
-                    selections.Add($"{structure.Values[index]} AS {quotedColName}");
-                    index++;
+                    selections.Add($"{paramName!} AS {quotedColName}");
```

Comment:
Do we need the null forgiving operator here after paramName? Since its true, isnt it guaranteed to be not null?

Reply:
yeah correct thanks.